File: Contents/InitialConsonantSearch/InitialConsonantSearch/Main/TEST/Cell/StandardLineSettingCell.m
//
//  StandardLineSettingCell.m
//  ToolSettingTest
//
//  Created by Kwan Hyun Son on 2023/09/18.
//

#import "StandardLineSettingCell.h"
#import <IosKit/IosKit.h>

@interface StandardLineSettingCell ()
@property (weak, nonatomic) IBOutlet UIButton *selectionBtn; // 단순한 라벨일 뿐이다. 하나 제외.
@property (weak, nonatomic) IBOutlet UIView *contentContainer;

@property (weak, nonatomic) IBOutlet MGUBrightnessBorderButton *colorBtn;
@property (weak, nonatomic) IBOutlet UIButton *ratioButton;
@property (weak, nonatomic) IBOutlet MGUDropdownButton *dropButton;
@property (weak, nonatomic) IBOutlet UISwitch *toggleSwitch;
@property (weak, nonatomic) IBOutlet UIButton *textInputButton;

@end

@implementation StandardLineSettingCell

- (void)awakeFromNib {
    [super awakeFromNib];
    if (self.selectionBtn != nil) {
        self.selectionBtn.titleLabel.adjustsFontSizeToFitWidth = YES;
        self.selectionBtn.titleLabel.minimumScaleFactor = 0.5;

        self.selectionBtn.tintAdjustmentMode = UIViewTintAdjustmentModeNormal;
        self.selectionBtn.userInteractionEnabled = NO;
        self.selectionBtn.translatesAutoresizingMaskIntoConstraints = NO;
        NSLayoutConstraint *constraint = [self.selectionBtn.heightAnchor constraintEqualToConstant:44.0]; // 키움: 55.0
        constraint.priority = UILayoutPriorityDefaultHigh;
        constraint.active = YES;
        [self.selectionBtn setTitleColor:[UIColor blackColor] forState:UIControlStateNormal];
    }
    CommonInit(self);
}

#pragma mark - 생성 & 소멸

static void CommonInit(StandardLineSettingCell *self) {
    self.contentContainer.backgroundColor = [UIColor clearColor];
    
    /// Top tableView
    if ([self.reuseIdentifier isEqualToString:StandardLineCellIDToggleDrop] == YES) {
        [self setupToggleDropMode];
    } else if ([self.reuseIdentifier isEqualToString:StandardLineCellIDToggle] == YES) {
        [self setupToggleMode];
    }
    
    /// Bottom tableView
    if ([self.reuseIdentifier isEqualToString:StandardLineCellIDNormal] == YES) {
        [self setupNormalMode];
    } else if ([self.reuseIdentifier isEqualToString:StandardLineCellIDUser] == YES) {
        [self setupUserMode];
    }
}

- (void)setupToggleDropMode {
    self.dropButton.backgroundColor = [UIColor clearColor];
    self.dropButton.defaultBackgroundColor = [UIColor whiteColor];
    self.dropButton.disabledBackgroundColor = [[UIColor lightGrayColor] colorWithAlphaComponent:0.2];
    self.dropButton.cellClass = [MGUTextDropdownCell class];
    self.dropButton.dropdownData = @[@"Left", @"Center", @"Right"];
    self.dropButton.selectedIndex = 0;
    self.dropButton.placeHolderTextAlignment = NSTextAlignmentCenter;
    self.dropButton.textAlignment = NSTextAlignmentLeft;
}

- (void)setupToggleMode {}

- (void)setupNormalMode {
    [self setupSelectionBtn];
    [self setupDropdownBtn];
}

- (void)setupUserMode {
    [self setupSelectionBtn];
    [self setupDropdownBtn];
    
    self.textField.autocorrectionType  = UITextAutocorrectionTypeNo;
    self.textField.spellCheckingType = UITextSpellCheckingTypeNo;
    self.textField.returnKeyType = UIReturnKeyDone;
    
    self.ratioButton.backgroundColor = [UIColor clearColor];
    self.ratioButton.layer.cornerRadius = 4.0; // 4.0
    
    self.textInputButton.backgroundColor = [UIColor whiteColor];
    self.textInputButton.layer.borderColor = [UIColor lightGrayColor].CGColor;
    self.textInputButton.layer.borderWidth = 1.0 / [UIScreen mainScreen].scale;
    self.textInputButton.layer.cornerRadius = 5.0;
    [self.textInputButton setTitleColor:[UIColor systemBlueColor] forState:UIControlStateNormal];
    [self.textInputButton setTitleColor:[UIColor grayColor] forState:UIControlStateDisabled];
    self.textInputButton.titleLabel.adjustsFontSizeToFitWidth = YES;
    self.textInputButton.titleLabel.minimumScaleFactor = 0.5;
}

/// Helper
- (void)setupSelectionBtn {
    self.selectionBtn.userInteractionEnabled = YES; // Label 형태로 사용. color fill 버튼 제외
    if (@available(iOS 13, *)) {
        if (@available(iOS 15, *)) {
            UIButtonConfiguration *configuration = self.selectionBtn.configuration;
            if (configuration != nil) {
                configuration.baseBackgroundColor = [UIColor clearColor];
                self.selectionBtn.configuration = configuration;
            }
        }
        // UIImageSymbolConfiguration *config = [UIImageSymbolConfiguration configurationWithScale:UIImageSymbolScaleMedium];
        // selectedImage = [UIImage systemImageNamed:@"checkmark.circle.fill" withConfiguration:config];
        // normalImage = [UIImage systemImageNamed:@"checkmark.circle" withConfiguration:config];
    } else {
        self.selectionBtn.imageView.contentMode = UIViewContentModeCenter;
        UIColor *tintColor = [UIColor colorWithRed:15.0/255.0 green:40.0/255.0 blue:71.0/255.0 alpha:1.0];
        self.selectionBtn.tintColor = tintColor;
        self.selectionBtn.imageView.tintColor = tintColor;
        // NSBundle *designTimeBundle = [NSBundle bundleForClass:[self classForCoder]];
        // selectedImage = [UIImage imageNamed:@"checkmark.circle.fill"
        //                                     inBundle:designTimeBundle
        //                compatibleWithTraitCollection:nil];
        // selectedImage = [selectedImage imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
        // normalImage = [UIImage imageNamed:@"uncheckmark.circle"
        //                                   inBundle:designTimeBundle
        //              compatibleWithTraitCollection:nil];
        // normalImage = [normalImage imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
    }
    NSBundle *designTimeBundle = [NSBundle bundleForClass:[self classForCoder]];
    UIImage *selectedImage = [UIImage imageNamed:@"checkmark.circle.fill"
                                        inBundle:designTimeBundle
                   compatibleWithTraitCollection:nil];
    selectedImage = [selectedImage imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
    UIImage *normalImage = [UIImage imageNamed:@"uncheckmark.circle"
                                      inBundle:designTimeBundle
                 compatibleWithTraitCollection:nil];
    normalImage = [normalImage imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
    [self.selectionBtn setImage:selectedImage forState:UIControlStateSelected];
    [self.selectionBtn setImage:normalImage forState:UIControlStateNormal];
    [self.selectionBtn setTitleColor:[UIColor blackColor] forState:UIControlStateSelected];
    [self.selectionBtn setTitleColor:[UIColor lightGrayColor] forState:UIControlStateNormal];
}

- (void)setupDropdownBtn {
    self.dropButton.backgroundColor = [UIColor clearColor];
    self.dropButton.cellClass = [MGULineWidthDropdownCell class];
    self.dropButton.dropdownData = @[@(-12), @(0), @(1), @(2) , @(3), @(4), @(5)];
    self.dropButton.selectedIndex = 0;
//    self.dropButton.placeHolderTextAlignment = NSTextAlignmentCenter;
//    self.dropButton.textAlignment = NSTextAlignmentLeft;
}


#pragma mark - 세터 & 게터

- (void)setData:(DTOStandardLineSetting *)model {
    _data = model;
    [self.selectionBtn setTitle:model.title forState:UIControlStateNormal];
    if ([self.reuseIdentifier isEqualToString:StandardLineCellIDToggleDrop] == YES) {
        NSLog(@"==> %ld %ld %@", self.dropButton.selectedIndex, model.dropBtnSelectedIndex, model);
        self.dropButton.selectedIndex = model.dropBtnSelectedIndex;
        
        if (self.toggleSwitch.isOn != model.isToggleOn) {
            self.toggleSwitch.on = model.isToggleOn;
        }
        if (self.dropButton.isEnabled != !model.dropBtnHidden) {
            self.dropButton.enabled = !model.dropBtnHidden;
        }
        //
        // hidden에서 disable로 바뀌었다.
        // if (self.dropButton.isHidden != model.dropBtnHidden) {
        //     self.dropButton.hidden = model.dropBtnHidden;
        // }
    } else if ([self.reuseIdentifier isEqualToString:StandardLineCellIDToggle] == YES) {
        if (self.toggleSwitch.isOn != model.isToggleOn) {
            self.toggleSwitch.on = model.isToggleOn;
        }
    } else if ([self.reuseIdentifier isEqualToString:StandardLineCellIDNormal] == YES) {
        if (self.selectionBtn.isSelected != model.selected) {
            self.selectionBtn.selected = model.selected;
        }
        self.dropButton.selectedIndex = model.dropBtnSelectedIndex;
        self.colorBtn.backgroundColor = model.color;
    } else if ([self.reuseIdentifier isEqualToString:StandardLineCellIDUser] == YES) {
        if (self.selectionBtn.isSelected != model.selected) {
            self.selectionBtn.selected = model.selected;
        }
        if (self.ratioButton.isSelected != model.isRatio) {
            self.ratioButton.selected = model.isRatio;
        }
        self.dropButton.selectedIndex = model.dropBtnSelectedIndex;
        self.colorBtn.backgroundColor = model.color;
        [self.textInputButton setTitle:model.textInputBtnTitle forState:UIControlStateNormal];
        self.textField.placeholder = model.textFieldPlaceHolderTitle;
        self.textField.text = model.textFieldTitle;
    }
}

#pragma mark - Actions

- (IBAction)switchValueChanged:(UISwitch *)sender {
    self.data.toggleOn = sender.isOn;
    self.data = self.data;
}

@end

File: Contents/InitialConsonantSearch/InitialConsonantSearch_Swift/Main/TEST/Cell/StandardLineSettingCell.swift
//
//  ChartSettingCell.swift
//  ChartType_Swift
//
//  Created by Kwan Hyun Son on 2023/09/08.
//

import UIKit
import IosKit

final class StandardLineSettingCell: UITableViewCell {
    
    // MARK: - Property
    
    weak var data: DTOStandardLineSetting? {
        didSet {
            guard let model = data
            else {
                return
            }
            selectionBtn.setTitle(model.title, for: .normal)
            if reuseIdentifier == StandardLineSettingCellID.toggleDrop.rawValue {
                dropButton.selectedIndex = model.dropBtnSelectedIndex
                if toggleSwitch.isOn != model.isToggleOn {
                    toggleSwitch.isOn = model.isToggleOn
                }
                if dropButton.isEnabled != !model.dropBtnHidden {
                    dropButton.isEnabled = !model.dropBtnHidden
                }
                //
                // hidden에서 disable로 바뀌었다.
                // if (self.dropButton.isHidden != model.dropBtnHidden) {
                //     self.dropButton.hidden = model.dropBtnHidden;
                // }
            } else if reuseIdentifier == StandardLineSettingCellID.toggle.rawValue {
                if toggleSwitch.isOn != model.isToggleOn {
                    toggleSwitch.isOn = model.isToggleOn
                }
                
            } else if reuseIdentifier == StandardLineSettingCellID.normal.rawValue {
                if selectionBtn.isSelected != model.isSelected {
                    selectionBtn.isSelected = model.isSelected
                }
                dropButton.selectedIndex = model.dropBtnSelectedIndex
                colorBtn.backgroundColor = model.color
            } else if reuseIdentifier == StandardLineSettingCellID.user.rawValue {
                if selectionBtn.isSelected != model.isSelected {
                    selectionBtn.isSelected = model.isSelected
                }
                if ratioButton.isSelected != model.isRatio {
                    ratioButton.isSelected = model.isRatio
                }
                dropButton.selectedIndex = model.dropBtnSelectedIndex
                colorBtn.backgroundColor = model.color
                textInputButton.setTitle(model.textInputBtnTitle, for: .normal)
                textField.placeholder = model.textFieldPlaceHolderTitle
                textField.text = model.textFieldTitle
            }
        }
    }
    
    @IBOutlet weak var textField: UITextField!
        
    @IBOutlet weak var selectionBtn: UIButton!
    @IBOutlet weak var contentContainer: UIView!
        
    @IBOutlet weak var colorBtn: SKUBrightnessBorderButton!
    @IBOutlet weak var ratioButton: UIButton!
    @IBOutlet weak var dropButton: SKUDropdownButton!
    @IBOutlet weak var toggleSwitch: UISwitch!
    @IBOutlet weak var textInputButton: UIButton!
    
    // MARK: - Override
    
    override func awakeFromNib() {
        super.awakeFromNib()
        if let selectionBtn = self.selectionBtn {
            selectionBtn.titleLabel?.adjustsFontSizeToFitWidth = true
            selectionBtn.titleLabel?.minimumScaleFactor = 0.5
            
            selectionBtn.tintAdjustmentMode = .normal
            selectionBtn.isUserInteractionEnabled = false
            selectionBtn.translatesAutoresizingMaskIntoConstraints = false
            let constraint = selectionBtn.heightAnchor.constraint(equalToConstant: 44.0)
            constraint.priority = .defaultHigh
            constraint.isActive = true
            selectionBtn.setTitleColor(.black, for: .normal)
        }
        commonInit()
    }
    
    // MARK: - 생성 & 소멸
    
    private func commonInit() {
        contentContainer.backgroundColor = .clear
           
        /// Top tableView
        if reuseIdentifier == StandardLineSettingCellID.toggleDrop.rawValue {
            setupToggleDropMode()
        } else if reuseIdentifier == StandardLineSettingCellID.toggle.rawValue {
            setupToggleMode()
        }
        
        /// Bottom tableView
        if reuseIdentifier == StandardLineSettingCellID.normal.rawValue {
            setupNormalMode()
        } else if reuseIdentifier == StandardLineSettingCellID.user.rawValue {
            setupUserMode()
        }
    }
    
    private func setupToggleDropMode() {
        dropButton.backgroundColor = .clear
        dropButton.defaultBackgroundColor = .white
        dropButton.disabledBackgroundColor = UIColor.lightGray.withAlphaComponent(0.2)
        dropButton.cellClass = SKUTextDropdownCell.self
        dropButton.dropdownData = ["Left", "Center", "Right"]
        dropButton.selectedIndex = 0
        dropButton.placeHolderTextAlignment = .center
        dropButton.textAlignment = .left
    }
    
    private func setupToggleMode() {
        // Implement setupToggleMode if needed
    }
    
    private func setupNormalMode() {
        setupSelectionBtn()
        setupDropdownBtn()
    }
    
    private func setupUserMode() {
        setupSelectionBtn()
        setupDropdownBtn()
            
        textField.autocorrectionType = .no
        textField.spellCheckingType = .no
        textField.returnKeyType = .done
            
        ratioButton.backgroundColor = .clear
        ratioButton.layer.cornerRadius = 4.0
            
        textInputButton.backgroundColor = .white
        textInputButton.layer.borderColor = UIColor.lightGray.cgColor
        textInputButton.layer.borderWidth = 1.0 / UIScreen.main.scale
        textInputButton.layer.cornerRadius = 5.0
        textInputButton.setTitleColor(.systemBlue, for: .normal)
        textInputButton.setTitleColor(.gray, for: .disabled)
        textInputButton.titleLabel?.adjustsFontSizeToFitWidth = true
        textInputButton.titleLabel?.minimumScaleFactor = 0.5
    }
    
    /// Helper
    private func setupSelectionBtn() {
        selectionBtn.isUserInteractionEnabled = true
        if #available(iOS 13, *) {
            if #available(iOS 15, *) {
                var configuration = selectionBtn.configuration
                configuration?.baseBackgroundColor = .clear
                selectionBtn.configuration = configuration
            }
            //
            // let config = UIImage.SymbolConfiguration(scale: .medium)
            // selectedImage = UIImage(systemName: "checkmark.circle.fill", withConfiguration: config)
            // normalImage = UIImage(systemName: "checkmark.circle", withConfiguration: config)
        } else {
            selectionBtn.imageView?.contentMode = .center
            let tintColor = UIColor(red: 15.0/255.0, green: 40.0/255.0, blue: 71.0/255.0, alpha: 1.0)
            selectionBtn.tintColor = tintColor
            selectionBtn.imageView?.tintColor = tintColor
            //
            // let designTimeBundle = Bundle(for: Self.self)
            // selectedImage = UIImage(named: "checkmark.circle.fill", in: designTimeBundle, compatibleWith: nil)
            // selectedImage = selectedImage?.withRenderingMode(.alwaysTemplate)
            // normalImage = UIImage(named: "checkmark.circle", in: designTimeBundle, compatibleWith: nil)
            // normalImage = normalImage?.withRenderingMode(.alwaysTemplate)
        }

        let designTimeBundle = Bundle(for: Self.self)
        var selectedImage = UIImage(named: "checkmark.circle.fill", in: designTimeBundle, compatibleWith: nil)
        selectedImage = selectedImage?.withRenderingMode(.alwaysTemplate)
        var normalImage = UIImage(named: "uncheckmark.circle", in: designTimeBundle, compatibleWith: nil)
        normalImage = normalImage?.withRenderingMode(.alwaysTemplate)
        selectionBtn.setImage(selectedImage, for: .selected)
        selectionBtn.setImage(normalImage, for: .normal)
        selectionBtn.setTitleColor(.black, for: .selected)
        selectionBtn.setTitleColor(.lightGray, for: .normal)
    }
        
    private func setupDropdownBtn() {
        dropButton.backgroundColor = .clear
        dropButton.cellClass = SKULineWidthDropdownCell.self
        dropButton.dropdownData = [-12, 0, 1, 2, 3, 4, 5]
        dropButton.selectedIndex = 0
    }
        
    // MARK: - Actions
    @IBAction func switchValueChanged(_ sender: UISwitch) {
        guard let data = data
        else {
            return
        }
        data.isToggleOn = sender.isOn
        self.data = data
    }
}
